### ISSUE TITLE
Enter event for drag'n'drop

### ISSUE THREAD

**Issue by emoon** [glfw/glfw#1898]
Something that would be useful is to have support for an event when a user begins and ends a dragging operation.
My use case is I want to be able to hint/show where the user can drop something, which isn't possible today as the current drop callback happens at the very end.

Just for reference this is what Qt supports https://doc.qt.io/qt-5/qwidget.html#dragEnterEvent which is 4 "events" (in practice 4 C++ methods one can implement)

```C++
void QWidget::dragEnterEvent(QDragEnterEvent *event)
```
```
This event handler is called when a drag is in progress and the mouse enters this widget. The event is passed in the event parameter.

If the event is ignored, the widget won't receive any drag move events.

See the Drag-and-drop documentation for an overview of how to provide drag-and-drop in your application.
```
```C++
void QWidget::dragLeaveEvent(QDragLeaveEvent *event)
```
```
This event handler is called when a drag is in progress and the mouse leaves this widget. The event is passed in the event parameter.

See the Drag-and-drop documentation for an overview of how to provide drag-and-drop in your application.
```
```C++
void QWidget::dragMoveEvent(QDragMoveEvent *event)
```
```
This event handler is called if a drag is in progress, and when any of the following conditions occur: the cursor enters this widget, the cursor moves within this widget, or a modifier key is pressed on the keyboard while this widget has the focus. The event is passed in the event parameter.

See the Drag-and-drop documentation for an overview of how to provide drag-and-drop in your application.
```
```C++
void QWidget::dropEvent(QDropEvent *event)
```
```
This event handler is called when the drag is dropped on this widget. The event is passed in the event parameter.

See the Drag-and-drop documentation for an overview of how to provide drag-and-drop in your application.
```

I'm not sure if glfw needs to support all of the above, but they are all useful when you need to implement drag`n`drop

**Comment by dougbinks** [comment]
Right, I think this would be via a window attribute similar to `GLFW_HOVERED`:

```
if (glfwGetWindowAttrib(window, GLFW_DRAGGING))
{
    highlight_drop_target();
}
```

This attribute might actually be the easiest additional API, as if it's set prior to the current cursor event callback it can be queried then without the need for other callbacks.

**Comment by Wesxdz** [comment]
Along with this, having drag and drop support from URLs in browsers would be great. X11 does call client messages but GLFW does not forward everything through GLFWdropfun

**Comment by dougbinks** [comment]
That is a separate issue, so I would open this as a new issue.

**Comment by Pilzschaf** [comment]
I have looked into this issue for myself a bit and tried to implement the GLFW_DRAGGING proposal. On X11 it works fine. But I think the Win32 backend might become a bit of a problem. As it seems to me, the only option to be notified about drag and drop upon window entering (and not only at drop release with the WM_DROPFILES message) seems to be the IDropTarget API. See https://docs.microsoft.com/en-us/windows/win32/api/oleidl/nn-oleidl-idroptarget. However I don't know if such an API is in the scope of GLFW.

**Comment by emoon** [comment]
I think Qt is using `IDropTarget` API on Win32 at least so it might be the only option.

**Comment by dougbinks** [comment]
GLFW is so far a C library, so using the standard C++ COM approach might be out of scope. However COM is a binary standard so can be used from C. I've been looking for Microsoft documentation, but can't find much, however this article explains the basics:

https://www.codeproject.com/Articles/13601/COM-in-plain-C

There is a minimal C++ example gist up at https://gist.github.com/mmozeiko/5b1700c558d37cb2bb2f00328c2982f6 which could also be used for reference.

I'd love to help more but am currently too busy.

**Comment by emoon** [comment]
Yeah, It's totally possible to using COM from C. I did it for D3D9 ages ago, Also the `COM-in-plain-C` article is a 8 part one which can be seen here https://www.codeproject.com/script/Articles/MemberArticles.aspx?amid=88625

In the case of D3D9 it was quite easy to use, but maybe MS provided some defines/wrappers for that case.

**Comment by falbrechtskirchinger** [comment]
Is anyone working on this issue? Otherwise, I may work on this myself.

**Comment by dougbinks** [comment]
Not that I am aware of.

**Comment by emoon** [comment]
Yeah, I thought about trying to take a stab at the COM stuff from C, but I never got around to do it so.

**Comment by Wesxdz** [comment]
I implemented it for x11 in an older version of glfw https://github.com/glfw/glfw/compare/master...Wesxdz:master if you want a reference

**Comment by falbrechtskirchinger** [comment]
I switched from Windows to Linux in 2008. I've worked with COM sometime before that. I'm not worried about X11 but will gladly take a look at previous work. Wayland and Cocoa will require more research, mostly Cocoa - I literally only know the name.

Once I have a draft PR, I'll need Windows and Mac OS users for testing.

**Comment by emoon** [comment]
I have all 3 OSes around for easy testing so I can help out with that.

**Comment by falbrechtskirchinger** [comment]
Thanks, everyone! I'm in the early planning stages for a project and not having this functionality is a showstopper. I won't work on this right away, but soon.

**Comment by falbrechtskirchinger** [comment]
A draft PR (#2128) is available. Just a very early `IDropTarget` implementation, but I'd appreciate any comments and feedback on the outline and questions in the description to inform my next steps.

**Comment by dougbinks** [comment]
Great. I've not had a chance to look at the code in detail, but for testing
and inclusing as a partial example I recommend adding code to
tests/events.c

**Comment by ws909** [comment]
I’m also available for Cocoa work and testing, but I’m new to the internals of GLFW. Have used Cocoa and Cocoa Touch from Swift in the past. I’m using the generated XCode project for development. It’s not great.

**Comment by raeleus** [comment]
I'm sorry to revive this, but I'm definitely interested in seeing this implemented for my apps. Thanks!

**Comment by tadeohepperle** [comment]
Is anyone still working on this? It would be great to be able to catch the drag enter and leave events in some way.

**Comment by Silverlan** [comment]
I'm also throwing my interest into the ring! I managed to work around it for windows, but x11 / wayland are a bit more tricky.

**Comment by emoon** [comment]
I'm considering looking into this now because I need it soon:ish. I just wonder if anyone else cc @Silverlan has made some more progress on it. Also before starting the implementation I know @dougbinks brought up some ideas / outlines for the API, but it would be great if we could try to narrow it down to what we want before I start doing it. I have access to wayland, macos and windows for the work (so if someone else could have a look at x11 that would be great)

**Comment by Wesxdz** [comment]
> I'm considering looking into this now because I need it soon:ish. I just wonder if anyone else cc [@Silverlan](https://github.com/Silverlan) has made some more progress on it. Also before starting the implementation I know [@dougbinks](https://github.com/dougbinks) brought up some ideas / outlines for the API, but it would be great if we could try to narrow it down to what we want before I start doing it. I have access to wayland, macos and windows for the work (so if someone else could have a look at x11 that would be great)

https://github.com/glfw/glfw/compare/master...Wesxdz:glfw:master

**Comment by Silverlan** [comment]
> I'm considering looking into this now because I need it soon:ish. I just wonder if anyone else cc [@Silverlan](https://github.com/Silverlan) has made some more progress on it. Also before starting the implementation I know [@dougbinks](https://github.com/dougbinks) brought up some ideas / outlines for the API, but it would be great if we could try to narrow it down to what we want before I start doing it. I have access to wayland, macos and windows for the work (so if someone else could have a look at x11 that would be great)

I'm using [Wesxdz](https://github.com/glfw/glfw/compare/master...Wesxdz:glfw:master) implementation (see post above), which works great with X11. For Wayland I have no solution either unfortunately.

For Windows I'm using the [IDropTarget interface](https://learn.microsoft.com/en-us/windows/win32/api/oleidl/nn-oleidl-idroptarget) (which has been suggested before):
https://github.com/Silverlan/iglfw/blob/master/src/implementation/file_drop_target.cppm#L19

**Comment by falbrechtskirchinger** [comment]
I still intend to finish this PR (https://github.com/glfw/glfw/pull/2128) for myself in H1/2026 (Q2 more likely than Q1), but I've yet to receive any feedback from @elmindreda. And besides some smaller design quirks I recall need fixing, I still need a decision on whether a more complete MIME type translation layer is welcome (I'm strongly in favor of this). Also, I'm not a Windows user, so I have no idea what may have changed API-wise in 4 years.

One thing that held me back was my inability to write and test a Cocoa implementation. I can read Objective-C++ but not effectively write it. LLMs should help with that today.

**Comment by dougbinks** [comment]
> One thing that held me back was my inability to write and test a Cocoa implementation. I can read Objective-C++ but not effectively write it. LLMs should help with that today.

Please do not use LLMs to code for GLFW. We can't accept copy-pasted code from other projects and so can't accept LLM code (which is essentially code copy pasted from the internet with some filtering).

> I still intend to finish this PR (https://github.com/glfw/glfw/pull/2128) for myself in H1/2026 (Q2 more likely than Q1), but I've yet to receive any feedback from @elmindreda. And besides some smaller design quirks I recall need fixing, I still need a decision on whether a more complete MIME type translation layer is welcome (I'm strongly in favor of this). Also, I'm not a Windows user, so I have no idea what may have changed API-wise in 4 years.

Your PR looks like it's trying to take on a lot more features than drag and drop enter/leave events. This makes it fairly complicated and slower to review and/or merge.

GLFW currently only supports one type of drag and drop, [path drop input](https://www.glfw.org/docs/latest/input_guide.html#path_drop). For simplicity an enter/leave callback API should be designed specifically for that.

**Comment by emoon** [comment]
I can help out with Wayland if no-one has that covered (I haven't really worked that much with Wayland, but that is what I'm running so I can try to figure it out)

**Comment by falbrechtskirchinger** [comment]
@dougbinks 
> > One thing that held me back was my inability to write and test a Cocoa implementation. I can read Objective-C++ but not effectively write it. LLMs should help with that today.
> 
> Please do not use LLMs to code for GLFW. We can't accept copy-pasted code from other projects and so can't accept LLM code (which is essentially code copy pasted from the internet with some filtering).

That's not what I said. I can read documentation and write pseudocode. An LLM can translate that into syntactically correct Objective-C++. I can't expect someone helping out with MacOS to go back and forth with me in a comment section, debugging syntax issues.
I got as far as reading the Cocoa drag&drop API documentation, but then failed at the first line of Objective-C++.
 
> Your PR looks like it's trying to take on a lot more features than drag and drop enter/leave events. This makes it fairly complicated and slower to review and/or merge.

The goal was to settle on an API that can handle most, if not all, scenarios (hence the [RFC] tag and draft status). A proof-of-concept can then be broken into smaller pieces and refined.

> GLFW currently only supports one type of drag and drop, [path drop input](https://www.glfw.org/docs/latest/input_guide.html#path_drop). For simplicity an enter/leave callback API should be designed specifically for that.

I haven't looked at my PR in years, but generally, I try to make common things easy. I'd be surprised if I hadn't thought of that.

<hr />

@emoon
> I can help out with Wayland if no-one has that covered (I haven't really worked that much with Wayland, but that is what I'm running so I can try to figure it out)

Thanks. I can cover Wayland, X11, and Windows (in a VM). I don't have access to MacOS, but @ws909 offered assistance.
And I thought I got drag&drop of text and paths working on those 3 platforms.

<hr />

If I can avoid my initial efforts going to waste, I'll do my best to prioritise this ASAP, but the earliest I can make serious time for this is next month.

**Comment by emoon** [comment]
Ah great! Well, I'm available to help out in case you need for both implementation and testing.
